benchmarks/distributed/rpc/parameter_server/servers/AverageParameterServer.py

```diff
@@ -0,0 +1,96 @@
+import threading
+
+import torch
+import torch.distributed.rpc as rpc
+
+from .AverageParameterServerBase import AverageParameterServerBase
+from .ParameterServerBase import ParameterServerBase
+
+
+class AverageParameterServer(AverageParameterServerBase):
+
+    def __init__(
+        self,
+        rank,
+        trainer_count,
+        use_cuda_rpc
+    ):
+        super().__init__(rank)
+
+        self.lock = threading.Lock()
+
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.use_cuda_rpc = use_cuda_rpc
+
+        self.batch_number = 0
+        self.futures = {}
+        self.gradient_dict = {}
+
+    @staticmethod
+    def reset_state(ps_rref):
+        self = ps_rref.local_value()
+        self.batch_number = 0
+        self.futures.clear()
+        self.gradient_dict.clear()
+        self.clear_metrics()
+
+    @staticmethod
+    def get_metrics_rpc(ps_rref):
+        self = ps_rref.local_value()
+        return self.get_metrics()
+
+    def param_key(self, param_loc):
+        return f"{self.batch_number},{param_loc}"
+
+    def clear_batch_state(self):
+        self.futures.clear()
+        self.gradient_dict.clear()
+
+    def process_gradient(self, gradient, param_loc):
```

Comment:
Would I be correct if I assume `param_loc` means `param_index`? Besides is the loc here mean the same thing as the loc in `BP_LOC_STRAGGLER`?

Reply:
> Would I be correct if I assume param_loc means param_index?

yes `param_loc` is zero indexed `param_index` if all the params are part of the same list.

> Besides is the loc here mean the same thing as the loc in BP_LOC_STRAGGLER?

yes need to update `BP_LOC_STRAGGLER`

I'll also update loc to index in the relevant variables.

Reply:
the parameter index isn't needed. I will use the gradient bucket index.